Add ArmorTypes to HitShape

reaperrr opened the pull request:
This allows to assign Armor traits to specific HitShapes, allowing to make certain parts of actors more vulnerable or durable than others.
Can also be used to simulate directional armor in a good-enough way that this closes #7718.

Testcase can be found here: https://github.com/reaperrr/OpenRA/tree/hitshape-armor-test

Closes #9268

reaperrr committed:
Add ArmorTypes to HitShape

This allows to assign Armor traits to specific HitShapes, allowing to make certain parts of actors more vulnerable or durable than others.
Can also be used to simulate directional armor.

obrakmann commented:
Of course this means that we would have to add the ability to drive in reverse to prevent units from showing the weaker rear armor when retreating.

ltem commented:
The testcase works great.

-------------------------

Following obrakmann's reverse-drive-ability note, a way to position units like  the Tiberium Wars Formation Move would also be useful.

|![image](https://user-images.githubusercontent.com/20945684/28499549-c2499a5c-6fb8-11e7-98c8-a91b17be324d.png)|
|----|

pchote commented:
I don't think we shoudl add directional armor or reverse move to our default mods, but I see no reason why other mods couldn't do both (aside from Mobile/Move being crap).

abcdefg30 commented on OpenRA.Mods.Common/Warheads/DamageWarhead.cs:
```diff
@@ -63,5 +74,27 @@ public virtual void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damag
+
+			// PERF: Avoid LINQ.
+			foreach (var at in armorTypes)
+				if (armor.StartsWith(at))
```
Why are you using `StartsWith`? This should also get a `StringComparison.Ordinal`.

reaperrr replied:
I wasn't sure how spaces are handled, nor what else to use instead.

abcdefg30 reviewed (dismissed):
Looks good to me otherwise.

reaperrr commented:
Updated.

reaperrr commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -25,6 +25,10 @@ public class HitShapeInfo : ConditionalTraitInfo, Requires<BodyOrientationInfo>
 
+		[Desc("Defines which Armor types apply when the actor receives damage to this HitShape.",
+			"If none specified, all armor types the actor has are valid.")]
+		public readonly HashSet<string> ArmorTypes = new HashSet<string>();
```
I'll see if I can come up with something, at least for TargetDamageWarhead.

reaperrr committed:
Add some summaries to DamageWarhead

reaperrr committed:
Add HitShape.ArmorTypes awareness to TargetDamageWarhead

reaperrr committed:
Make HealthPercentageDamageWarhead HitShape-trait-aware

Note: This does not yet take the actual 'shape' into consideration (it takes the shape closest to/centered around the actor.CenterPosition), but it now
- doesn't deal damage when the actor has no enabled HitShape trait
- takes HitShape.ArmorTypes into consideration

reaperrr commented:
Updated. I haven't had time to test the Target/HealthPercentageWH changes yet, though.

RoosterDragon commented on OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs:
```diff
@@ -43,12 +43,20 @@ public override void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> dama
+			if (!activeShapes.Any())
+				return;
+
+			var closestActiveShape = activeShapes.MinBy(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
```
One pass option:
```c#
var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
    .MinByOrDefault(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
if (closestActiveShape == null)
    return;
```
Requires you to allocate the MinBy lambda even if there are no active hit shapes, but I'm assuming that's the uncommon case/a worthwhile tradeoff.

reaperrr commented:
Updated.

reaperrr commented:
@OpenRA/engine-hackers This is basically blocking some related follow-ups, so it would be nice if this could get another review soon.

pchote reviewed (commented):
I really want to be able to :+1: this, but i'm sorry to say that this really is still setting off my future-maintenance-nightmare spidey-sense.

The core of the problem is that this introduces a second channel for applying damage (DoImpact with and without a hitshape) right in the middle of a chain of inheritance and overrides.  This makes it very difficult to know which will be used, and means that the same weapon/warheads/vs fired by different means will end up doing different amounts of damage to the same actor.

This feature really needs to be integrated from the bottom up, and not just bolted on the side.  Require *all* damaging warheads to specify the hitshape, and fix the higher level code to be able to choose the appropriate one based on whatever rules make sense in each case.

pchote commented:
There isn't a clear way forwards here, so moving to the Future milestone.
